### ISSUE TITLE
[MdTable] Initialization error in data fetching

### ISSUE THREAD

**Issue by lucacattide** [vuematerial/vue-material#1870]
### Steps to reproduce

- Use ```MdTable``` component including full features (custom template, selection, search, multiple selection, sorting, etc.);
- Initialize it with via [_Vue Router_ data fecthing post-navigation](https://router.vuejs.org/guide/advanced/data-fetching.html#fetching-after-navigation) technique with _axios_ (tried with pre-navigation too with the same result). Data are fetched from a _MySQL_ DB through _Laravel_ API;

### Which browser?

- Chrome

### What is expected?

- ```MdTable``` correctly rendered with data fetched

### What is actually happening?

Hi everyone.

As described above, table doesn't been rendered due to a ```null/empty v-model```, as may be deducted from the scenario. When the component is rendered, an exception beign thrown:

> Cannot read property 'id' of null

I'm using two variables to work with it. ```elemento``` is being used for getting data and use them in various context. ```cercato``` is being used as table model, searching feature and so on.
Table rows are managed with custom template, like this:
```
<!--Inizio Elementi-->
<md-table-row slot="md-table-row" slot-scope="{ item }" :class="getClass(item)" :md-selectable="selezioneTipo" @click="visualizzaElemento(item.id)"
    md-auto-select>
    <md-table-cell md-label="ID" md-sort-by="id" md-numeric>{{ item.id }}</md-table-cell>
    <md-table-cell v-for="(val, i) in item" :key="val.id" v-if="(i !== 'id') && (i !== 'estensione_garanzia') && (i !== 'note')"
        :md-label="formattaHeader(i)" :md-sort-by="i">
        <span v-if="(i === 'labeled') && (val === true)">Si</span>
        <span v-else-if="(i === 'labeled') && (val === false)">No</span>
        <span v-else-if="(i === 'stato') && (val === true)">In Uso</span>
        <span v-else-if="(i === 'stato') && (val === false)">Dismesso</span>
        <span v-else-if="(i === 'data_dismissione') && (val === '')">-</span>
        <span v-else>{{ val }}</span>
    </md-table-cell>
</md-table-row>
<!--Fine Elementi-->
```

With this data fetching approach:

```
data: () => ({
    cercato: [],
    elemento: [],
    caricamento: false
},
watch: {
    '$route' : 'setElemento'
},
methods: {
    setElemento() {
        this.elemento = null
        // Lodaing...
        this.caricamento = true

        // Data Fetching
        axios.get(`/${this.$route.meta.description}`)
        .then(elemento => {
            // Initialize data
            this.elemento = elemento.data.data
            // Initialize table data
            this.cercato = this.elemento
        })
        .catch(errore => this.$emit('errore', errore))
        .finally(() => this.caricamento = false)
    }
},
created() {
    this.setElemento()
}
```

I noticed that the model variable beign initialized correctly, as shown here after ajax call:

![screen shot 2018-07-23 at 10 24 55](https://user-images.githubusercontent.com/12085296/43066370-c38f226c-8e64-11e8-9e37-1f71fe73f1eb.png)
![screen shot 2018-07-23 at 10 25 51](https://user-images.githubusercontent.com/12085296/43066373-c5acaf10-8e64-11e8-876d-e19abbb01ba2.png)

There's a way to render them correctly? I tried to use ```$forceUpdate``` too but ```MdTable``` not being rendered neither with that.

Thanks in advance for help.

**Comment by Samuell1** [comment]
Hey, do you tried to not set null but empty array?

**Comment by lucacattide** [comment]
Hi @Samuell1 , do you mean ```cercato```? if that so, as you can see from above code it's declared and initialized from scratch as empty array:

```
data: () => ({
    cercato: [],
    elemento: [],
    // ...
```

In fact, ```cercato``` initial state is rendered like this:

![screen shot 2018-07-23 at 12 32 37](https://user-images.githubusercontent.com/12085296/43071930-9915906a-8e74-11e8-8ab1-caea761665e5.png)

**Comment by Samuell1** [comment]
You can try to rerender table with v-if on md-table element

**Comment by lucacattide** [comment]
After many many days of analysis and tries, I cornered the source of issue to the ```slot-scope``` ```md-table-row``` attribute. So, actually the slotted version of ```MdTable``` crashes on data fetching initialization. It should be extended to treat remote data as I suggested above.
For now, I refactored this section in my project with a simply polyfill - terrible approach, please solve this on the next future releases:
```
<md-table-row :class="getClass(val)" :md-selectable="selezioneTipo" @click="visualizzaElemento(val.id)" v-for="(val) in cercato"
    :key="val.id">
    <md-table-cell md-label="ID" md-sort-by="id" md-numeric>{{ val.id }}</md-table-cell>
    <md-table-cell v-for="(valore, j) in val" :key="j" v-if="(j !== 'id') && (j !== 'estensione_garanzia') && (j !== 'note')"
        :md-label="formattaHeader(j)" :md-sort-by="j">
        <span v-if="(j === 'labeled') && (valore === 1)">Si</span>
        <span v-else-if="(j === 'labeled') && (valore === 0)">No</span>
        <span v-else-if="(j === 'stato') && (valore === '1')">In Uso</span>
        <span v-else-if="(j === 'stato') && (valore === '0')">Dismesso</span>
        <span v-else-if="((j === 'data_acquisto') || (j === 'data_dismissione') || (j === 'parent_id')) && (valore === null)">-</span>
        <span v-else-if="((j === 'data_acquisto') || (j === 'data_dismissione')) && (valore !== null)">{{
            formattaData(valore.date) }}</span>
        <span v-else>{{ valore }}</span>
    </md-table-cell>
</md-table-row>
```
By leaving apart the ```spans``` - that's conditional stuff from my project - I simply omitted ```slot="md-table-row"``` and ```slot-scope="{ item }"``` from ```md-table-row``` tag, delegating the iteration to a raw ```v-for``` construct. the nested ```v-for``` manage respectively data rows and columns. Selection is still usable by declaring```val``` as ```getClass``` parameter.

This workaround presents some issues:
- Single selection: on row click the element being correctly selected but the UX goes crazy (selected element CSS class being triggered on all elements);
- Multiple selection: doesn't properly work (still working as single-element);
- Search: empty state on zero items returned not being triggered;

That's all for now. I'll stay tuned for any fixing news or better solution.

**Comment by lucacattide** [comment]
After the 300th time that I revisioned this code, I found it.

On this line:
```
[...]
<md-table-cell v-for="(val, i) in item" :key="val.id" v-if="(i !== 'id') && (i !== 'estensione_garanzia') && (i !== 'note')"
[...]
```
this is the guilty syntax:
```
:key="val.id"
```
This funny syntax was responsible for component broken initialization. Of course there's no ```id``` property in data structure (array of objects without nested objects, as from official doc). By changing it to:
```
:key="i"
```
Everything works. So I confirm that ```md-table``` officially supports data fetching with no issues.

Thanks everyone for patience. This lesson teached me that sometimes it's better to switch off and take a break from everything :)
